src/Orleans.Runtime/Services/GrainServiceFactory.cs

```diff
@@ -0,0 +1,35 @@
+using Orleans.Services;
+
+namespace Orleans.Runtime
+{
+    public interface IGrainServiceFactory
+    {
+        /// <summary>
+        /// Creates a grain reference for a grain service instance on a given silo.
+        /// This is used by grain indexing.
+        /// </summary>
+        GrainReference MakeGrainServiceReference(int typeData, string systemGrainId, SiloAddress siloAddress);
```

Comment:
s/typeData/typeCode... but still awkward since we don't expose that info via other APIs